### DISCUSSION TITLE
How to overcome Docker Hub download rate limit with Apicurio registry installed via OperatorHub?

### DISCUSSION THREAD

**Discussion by mzaikin** [Apicurio/apicurio-registry#2014] [Q&A]
Disclaimer: this is not a duplicate of the question I asked in the Issues section (https://github.com/Apicurio/apicurio-registry/issues/2009).

Current setup: Apicurio Registry Operator 0.0.4-v1.3.2.final installed in OpenShift cluster, use JPA as persistence.

I wonder if OpenShift is starting experiencing download rate limitation of Apicurio Registry image by Docker Hub, and pulls fail, and lots of pods in the crashed state, and editing of ApicurioRegistry's `spec.image.name` not switching to new [locally pulled] `apicurio-registry-jpa` image, even while operator's description says different:

```
 description: >-
              Override the image name according to selected Persistence Type for
              example,
              MEM="docker.io/apicurio/apicurio-registry-mem:1.3.2.Final"
              KAFKA="docker.io/apicurio/apicurio-registry-kafka:1.3.2.Final"
              STREAMS="docker.io/apicurio/apicurio-registry-streams:1.3.2.Final"
              JPA="docker.io/apicurio/apicurio-registry-jpa:1.3.2.Final"
              INFINISPAN="docker.io/apicurio/apicurio-registry-infinispan:1.3.2.Final"
```

I wonder how to overcome it?

Delete Apicurio Registry Operator and recreate ApicurioRegistry, providing from the beginning new local Docker image?

If yes, how fully clean up all objects related to the existing Apicurio Registry from OpenShift?

Is there any chance to keep old schemas we populated, i.e. keep Postgre, and when recreating registry, provide currently existing DB?

Appreciate any help or hint.

**Comment by Ganesh-403** [comment]
@mzaikin The Docker Hub rate limit itself shouldn't require you to recreate your Registry or lose your data.

Since you're using the **JPA persistence** variant, your schemas are stored in PostgreSQL, not in the Registry container. As long as you keep the same database, you can redeploy the Registry and it will continue using the existing artifacts. The Registry is effectively stateless with respect to schema storage. :contentReference[oaicite:0]{index=0}

Regarding the image override:

```yaml
spec:
  image:
    name: ...
```

the Operator documentation indicates that the image can be overridden, for example:

```text
docker.io/apicurio/apicurio-registry-jpa:1.3.2.Final
```

So pointing it to another registry (or an internally mirrored image) is the intended approach. If changing `spec.image.name` isn't taking effect, that sounds more like an issue with the Operator version (`0.0.4-v1.3.2.final`) than with the Registry itself. :contentReference[oaicite:1]{index=1}

If your goal is simply to avoid Docker Hub rate limits, I'd recommend:

1. Mirror the required `apicurio-registry-jpa` image into your internal registry (or another registry you control).
2. Update the `spec.image.name` to reference the mirrored image.
3. Keep using the existing PostgreSQL database.
4. Recreate only the Registry deployment if necessary—the database can remain untouched.

There should be **no need to delete PostgreSQL** or re-import your schemas, provided the new Registry instance is configured to connect to the same database.

A few questions that would help determine why the override isn't working:

1. Which OpenShift version are you running?
2. After editing `spec.image.name`, does the `ApicurioRegistry` custom resource retain your change, or does the Operator immediately reconcile it back?
3. Are the failing pod events showing that they're still trying to pull `docker.io/apicurio/apicurio-registry-jpa:1.3.2.Final`, or are they attempting to pull your mirrored image?

If the Operator continues to ignore `spec.image.name`, that would suggest an Operator-specific bug or limitation rather than an issue with the Registry itself.

If this solves your problem, feel free to mark it as the accepted answer so others can find it easily.
